crates/bevy_reflect/src/func/registry.rs

```diff
@@ -0,0 +1,221 @@
+use alloc::borrow::Cow;
+use core::fmt::Debug;
+use std::sync::{Arc, PoisonError, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use bevy_utils::HashMap;
+
+use crate::func::{DynamicFunction, IntoFunction};
+
+/// A registry of [reflected functions].
+///
+/// This is the function-equivalent to the [`TypeRegistry`].
+///
+/// [reflected functions]: crate::func
+/// [`TypeRegistry`]: crate::TypeRegistry
+#[derive(Default)]
+pub struct FunctionRegistry {
+    /// Maps function [names] to their respective [`DynamicFunctions`].
+    ///
+    /// [names]: DynamicFunction::name
+    /// [`DynamicFunctions`]: DynamicFunction
+    functions: HashMap<Cow<'static, str>, DynamicFunction<'static>>,
+}
+
+impl FunctionRegistry {
+    /// Attempts to register the given function if it has not yet been registered already.
+    ///
+    /// Functions are mapped according to their [name].
+    /// If a `DynamicFunction` with the same name already exists, it will not be registered again.
+    /// To register the function anyway, overwriting any existing registration, use [`overwrite_registration`] instead.
+    ///
+    /// [name]: DynamicFunction::name
+    /// [`overwrite_registration`]: Self::overwrite_registration
+    pub fn register<F, Marker>(&mut self, function: F)
```

Comment:
Should we return a result? Unlike `TypeRegistry::register` we can't be certain a registration will succeed (i.e. it has the same name as an existing function).

Additionally, if did return a result, we could have an error like `FunctionRegistrationError::UnnamedClosure` to handle closures without a name, since they'll always be given a default name of `path::to::my::{{closure}}`. We don't _have_ to do that, but it might help prevent users from accidentally colliding with another closure defined in the same scope.

Reply:
Yes, I think it would be better if this returned a result which we unwrap in the app builder function.